File: src/main/kotlin/ru/digital/image/picture/images/ImageColorChange.java
package ru.digital.image.picture.images;

import com.example.demo.images.ArrayPointColorGetter;
import com.example.demo.images.OnePointColorGetter;
import com.example.demo.images.PointColorGetter;

import java.awt.image.BufferedImage;
import java.util.*;

public class ImageColorChange {

    private boolean hasTwoPointsColor;
    List<PointColorGetter> pointColorGetters;
    public ImageColorChange(List<PointColorGetter> pointColorGetters){
        this.pointColorGetters = pointColorGetters;
    }
    public void run(BufferedImage image){
        for(int y = 0; y < image.getHeight(); y++) {
            for (int x = 0; x < image.getWidth(); x++) {
                int rgb = image.getRGB(x,y);
                for(PointColorGetter getter : pointColorGetters){
                    if(getter instanceof OnePointColorGetter){
                        rgb = ((OnePointColorGetter)getter).get(image.getRGB(x,y));
                    }
                    if(getter instanceof ArrayPointColorGetter){
                        if(x < image.getWidth() -1 && y < image.getHeight() -1)
                            rgb = ((ArrayPointColorGetter)getter).get(
                                    new int[]{
                                    image.getRGB(x,y),
                                            applyBefore(image, x,y+1, getter),
                                            applyBefore(image, x+1,y, getter)});
                    }
                }
                image.setRGB(x, y, rgb);
            }
        }
    }

    private int applyBefore(BufferedImage image, int x,int y, PointColorGetter pointColorGetter){
        int rgb  = image.getRGB(x,y);
        for(PointColorGetter getter : pointColorGetters){
            if(getter == pointColorGetter){
                break;
            }

            if(getter instanceof OnePointColorGetter){
                rgb = ((OnePointColorGetter)getter).get(rgb);
            }
            if(getter instanceof ArrayPointColorGetter){
                if(x < image.getWidth() -1 && y > image.getHeight() -1)
                    rgb = ((ArrayPointColorGetter)getter).get(
                            new int[]{
                                    rgb,
                                    applyBefore(image, x,y+1, getter),
                                    applyBefore(image, x+1,y, getter)});
            }
        }
        return rgb;
    }
}

File: src/main/kotlin/ru/digital/image/picture/images/changer/ImageColorChange.kt
package ru.digital.image.picture.images.changer

import java.awt.image.BufferedImage

class ImageColorChange(
    val pointColorGetters: List<PointColorGetter>?
) {

    fun run(image: BufferedImage) {
        for (y in 0 until image.height) {
            for (x in 0 until image.width) {
                var rgb = image.getRGB(x, y)
                for (getter in pointColorGetters!!) {
                    if (getter is OnePointColorGetter) {
                        rgb = getter[image.getRGB(x, y)]
                    }
                    if (getter is ArrayPointColorGetter) {
                        if (x < image.width - 1 && y < image.height - 1) rgb = getter[intArrayOf(
                            image.getRGB(x, y),
                            applyBefore(image, x, y + 1, getter),
                            applyBefore(image, x + 1, y, getter)
                        )]
                    }
                }
                image.setRGB(x, y, rgb)
            }
        }
    }

    private fun applyBefore(image: BufferedImage, x: Int, y: Int, pointColorGetter: PointColorGetter): Int {
        var rgb = image.getRGB(x, y)
        for (getter in pointColorGetters!!) {
            if (getter === pointColorGetter) {
                break
            }
            if (getter is OnePointColorGetter) {
                rgb = getter[rgb]
            }
            if (getter is ArrayPointColorGetter) {
                if (x < image.width - 1 && y > image.height - 1) rgb = getter[intArrayOf(
                    rgb,
                    applyBefore(image, x, y + 1, getter),
                    applyBefore(image, x + 1, y, getter)
                )]
            }
        }
        return rgb
    }
}
